File: src/Builder/Java/origin/Robot.java
package Builder.Java.origin;


/**
 * Created by prefert on 2017/7/16.
 * 原始写法
 */
public class Robot {
    private String code;
    private String name;
    private int type;
    private int battery;
    private int ability;
    private double weight;
    private double height;

    // 通常我们会生成一个含有全部参数的构造函数
    public Robot(String code, String name, int type, int battery, int ability, double weight, double height) {
        this.code = code;
        this.name = name;
        this.type = type;
        this.battery = battery;
        this.ability = ability;
        this.weight = weight;
        this.height = height;
    }

    public Robot(String code, String name, int type) {
        this.code = code;
        this.name = name;
        this.type = type;
    }

    @Override
    public String toString() {
        return "Robot {" +
                "code = " + code + '\'' +
                ", name = '" + name + '\'' +
                ", type = '" + type + '\'' +
                ", battery = " + battery +
                ", ability = " + ability +
                ", weight = '" + weight + '\'' +
                ", height = '" + height + '\'' +
                '}';
    }
}

File: src/Builder/Scala/simulateJava/Robot.scala
package Builder.Scala.simulateJava

/**
  * Created by prefert on 2017/7/16.
  * 仿 java 版
  */
class Robot(builder: RobotBuilder) {
  val code = builder.code
  val name = builder.name
  val battery = builder.battery
}

class RobotBuilder {
  var code = ""
  var name = ""
  var battery = 0

  def setName(name: String): RobotBuilder = {
    this.name = name
    this // 返回 this 链式调用
  }

  def setCode(code: String): RobotBuilder = {
    this.code = code
    this
  }

  def setBattery(battery: Int): RobotBuilder = {
    this.battery = battery
    this
  }

  def build() = {
    new Robot(this)
  }
}



